Add verifications during DM deploy on subclouds

enzocandotti opened the pull request:
As part of improving the error reporting on subcloud deployment and upgrading, this change adds new  verifications after apply deploy_config file on subclouds.

After deploy_config is applied, the playbook
will wait until the unlock task is triggered.
When unlock task is triggered after applying the deploy-config file, is highly probable that the configuration applied is correct.

If unlock task is not triggered after some time,
it will check for unreconciled resources and errors into deployment pod logs. Deployment playbook will fail showing this information.

The output from this verifications will be catched by 'dcmanager subcloud errors' new command.

Test Plan:
PASS: Edit datanetwork type in config file in order to fail
      during apply. Verify that the error is catched by command.
PASS: Remove deployment namespace. Verify playbook fails waiting
      for pods to be ready. Verify dcmanager error command shows the
      error.
PASS: Edit controller-profile port names to invalid ones.
      Verify that apply is ok but the playbook fails due to timeout.
      Verify dcmanager error command shows the desired information.
PASS: Dcmanager subcloud reconfig using correct configuration.
      Verify that apply is ok and unlock is triggered.
      Verify that the playbook ends without failures.
PASS: Install DM on a SX system (region_name = RegionOne) with these changes.
      Verify that the new block is skipped.

Signed-off-by: Enzo Candotti <enzo.candotti@windriver.com>

yjian118 commented on docs/playbooks/wind-river-cloud-platform-deployment-manager.yaml:
```diff
@@ -14,8 +14,27 @@
+            source /etc/platform/openrc; hostname
+          register: get_host_name 
+
+        - name: Get region name
```
The subcloud's region name is not necessarily "subcloud*",  if you want to check if the host is a subcloud's node, better to check with"      shell: source /etc/platform/platform.conf; echo $distributed_cloud_role"

FabrizioPerez replied:
Makes sense. Testing the change. Thanks Yuxing!

enzocandotti replied:
Done.

yjian118 commented on docs/playbooks/wind-river-cloud-platform-deployment-manager.yaml:
```diff
@@ -284,3 +303,73 @@
+          register: fail_dm_logs
+          when: ("Unlocking" not in get_show_task_status.stdout)
+
+      when: (deploy_config is defined and "subcloud" in get_region_name.stdout)
```
The subcloud's region name is not reliable to check its dc role

enzocandotti replied:
Done. Thanks Yuxing!

yjian118 reviewed (changes requested):
I have some questions: 
1 .After this change, the deploy_status will be "deploying" before the controller-0 of the subcloud is unlocked, which means the deploy will take much longer time before being done, is this change of behavior expected? 
2. How about a multi-node subcloud, e.g. dx or standard subcloud?  If we want to think the subcloud is deployed after the controller-0 is unlocked and online, and the other nodes don't matter?

I'll need to add Matt as the core reviewer as this commit will change some logic

FabrizioPerez commented:
> I have some questions: 1 .After this change, the deploy_status will be "deploying" before the controller-0 of the subcloud is unlocked, which means the deploy will take much longer time before being done, is this change of behavior expected? 2. How about a multi-node subcloud, e.g. dx or standard subcloud? If we want to think the subcloud is deployed after the controller-0 is unlocked and online, and the other nodes don't matter?
> 
> I'll need to add Matt as the core reviewer as this commit will change some logic

For the first question, I followed this approach based on some requirements (or boundaries). For example, I was asked for not to run another playbook, instead add these verifications at the end of this one. Is a great question, but I can't see another solution taking this into account. Anyways, I'll appreciate suggestions.
Regarding the second question, I was asked to focus on Subclouds SX on this first stage. I think that multi-node could be an improvement on next steps.

enzocandotti commented:
Hi all,

Just to clarify, the idea of these changes is to wait until the task "Unlocking" is triggered after applying a deploy-config file. So, the playbook will not wait for the controller-0 to be back after the reboot, but it will be finished just before the unlock is started.

Doing this, we can gather logs from those errors that avoid the controller-0 to be unlocked. Due to some limitations, the action of waiting until the controller-0 is rebooted will require much more effort than expected, so we are not covering this on the current implementation.

On local tests, when there are no errors during deployment, the deployment playbook was extended no more than 1 minute.

On the following comment I will paste out the sample of a playbook output when deploying a AIO-DX subcloud including these changes.

enzocandotti commented:
AIO-Duplex Subcloud test:


subcloud duplex tested on 3 cases:

- Deployment with no errors
- Deployment with controller-0 error
- Deployment with controller-1 error

Steps to test test deployment of duplex subcloud using last changes:

- Create and bootstrap duplex subcloud.
- Create a dm-config for both nodes. Update and upload this file into controller-0 central to use this version.
- Deploy subcloud using reconfigure subcloud command (since subcloud is already bootstraped). Example:
dcmanager subcloud reconfig --sysadmin-password Li69nux* --deploy-config /home/sysadmin/config/subc/dm-config.yaml subcloud1


system:
[sysadmin@controller-0 ~(keystone_admin)]$ system show
+------------------------+--------------------------------------+
| Property               | Value                                |
+------------------------+--------------------------------------+
| contact                | None                                 |
| created_at             | 2022-10-31T11:27:24.525302+00:00     |
| description            | VDM Virtual Box AIO-DX System        |
| distributed_cloud_role | subcloud                             |
| https_enabled          | False                                |
| latitude               | None                                 |
| location               | vbox                                 |
| longitude              | None                                 |
| name                   | vbox                                 |
| region_name            | subcloud1                            |
| sdn_enabled            | False                                |
| security_feature       | spectre_meltdown_v1                  |
| service_project_name   | services                             |
| shared_services        | []                                   |
| software_version       | 22.12                                |
| system_mode            | duplex                               |
| system_type            | All-in-one                           |
| timezone               | UTC                                  |
| updated_at             | 2022-10-31T11:57:56.094743+00:00     |
| uuid                   | 222f0fd7-bede-4936-a72c-d32d8bca2984 |
| vswitch_type           | none                                 |
+------------------------+--------------------------------------+


---------------------------------------------------------------------------------------------------

*Case 1: Configuration ok

-output:

[sysadmin@controller-0 ~(keystone_admin)]$ dcmanager subcloud list 
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| id | name      | management | availability | deploy status | sync    | backup status | backup datetime |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| 11 | subcloud1 | unmanaged  | offline      | complete      | unknown | None          | None            |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
[sysadmin@controller-0 ~(keystone_admin)]$ dcmanager subcloud errors 11
No errors present


- /var/log/dcmanager/dcmanager.log:

2022-09-27 23:04:03.492 4055365 INFO dcmanager.manager.service [req-e437bfa7-70a3-4c27-8d31-676371346548 f0486ee592ab4e02bf039cd78048629e - - default default] Handling reconfigure_subcloud request for: 11
2022-09-27 23:04:03.493 4055365 INFO dcmanager.manager.subcloud_manager [req-e437bfa7-70a3-4c27-8d31-676371346548 f0486ee592ab4e02bf039cd78048629e - - default default] Reconfiguring subcloud 11.
2022-09-27 23:04:03.562 4055365 INFO dcmanager.manager.subcloud_manager [-] Starting deploy of subcloud1
2022-09-27 23:06:26.924 4055365 INFO dcmanager.manager.subcloud_manager [-] Successfully deployed subcloud1

- Ansible-playbook final:

TASK [Show waiting unlock] *****************************************************
Tuesday 27 September 2022  23:06:25 +0000 (0:01:33.107)       0:02:20.156 ***** 
ok: [subcloud1] => 
  msg:
  - 'waiting: Unlocking'
  - 'waiting: '

TASK [Retrieve kubectl resources reconciled status] ****************************
Tuesday 27 September 2022  23:06:25 +0000 (0:00:00.056)       0:02:20.212 ***** 
changed: [subcloud1]

TASK [Show unreconciled resources] *********************************************
Tuesday 27 September 2022  23:06:25 +0000 (0:00:00.426)       0:02:20.639 ***** 
ok: [subcloud1] => 
  msg:
  - 'recon: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false'

TASK [Get dm pod name] *********************************************************
Tuesday 27 September 2022  23:06:25 +0000 (0:00:00.056)       0:02:20.695 ***** 
changed: [subcloud1]

TASK [Retrieve dm logs] ********************************************************
Tuesday 27 September 2022  23:06:26 +0000 (0:00:00.500)       0:02:21.195 ***** 
changed: [subcloud1]

TASK [Show logs] ***************************************************************
Tuesday 27 September 2022  23:06:26 +0000 (0:00:00.457)       0:02:21.653 ***** 
ok: [subcloud1] => 
  msg:
  - 'Pod log: '
  - 'err: '

TASK [Waiting for trigger unlock.] **********************************************************
Tuesday 27 September 2022  23:06:26 +0000 (0:00:00.051)       0:02:21.704 ***** 
skipping: [subcloud1]

PLAY RECAP *********************************************************************
subcloud1                  : ok=41   changed=26   unreachable=0    failed=0   

Tuesday 27 September 2022  23:06:26 +0000 (0:00:00.024)       0:02:21.729 ***** 

---------------------------------------------------------------------------------------------------


* Case 2: Error in controller-0

In this case, procedure to test it is the same, but editing interface name
into controller0 hostprofile.
After that, run reconfigure command.

- Output:
[sysadmin@controller-0 subc_dx(keystone_admin)]$ dcmanager subcloud list
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| id | name      | management | availability | deploy status | sync    | backup status | backup datetime |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| 11 | subcloud1 | unmanaged  | offline      | deploy-failed | unknown | None          | None            |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+

[sysadmin@controller-0 subc_dx(keystone_admin)]$ dcmanager subcloud errors 11
FAILED deploying playbook of (subcloud1).
 detail: fatal: [subcloud1]: FAILED! => changed=false 
  msg:
  - Fail waiting for host unlock to be triggered. It could be due to DM config errors
  - "UNRECONCILED resources: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false"
  - "2022-10-31T11:58:14.047077043Z\tERROR\tcontroller.host\tuser data error\t{\"request\": \"deployment/controller-0\", \"error\": \"unable to find interface UUID for port: enp0s4\"}" 
FAILED TASK: TASK [Waiting for trigger unlock.]  Wednesday 28 September 2022  02:45:13 +0000 (0:00:00.065)       0:03:28.828


- /var/log/dcmanager/dcmanager.log:

2022-09-28 02:41:43.101 1550496 INFO dcmanager.manager.service [req-b7a081b6-bd8c-4347-8671-8590b52c2d21 f0486ee592ab4e02bf039cd78048629e - - default default] Handling reconfigure_subcloud request for: 11
2022-09-28 02:41:43.102 1550496 INFO dcmanager.manager.subcloud_manager [req-b7a081b6-bd8c-4347-8671-8590b52c2d21 f0486ee592ab4e02bf039cd78048629e - - default default] Reconfiguring subcloud 11.
2022-09-28 02:41:43.165 1550496 INFO dcmanager.manager.subcloud_manager [-] Starting deploy of subcloud1
2022-09-28 02:45:13.389 1550496 ERROR dcmanager.manager.subcloud_manager [-] FAILED deploying playbook of (subcloud1).
 detail: fatal: [subcloud1]: FAILED! => changed=false 
  msg:
  - Fail waiting for host unlock to be triggered. It could be due to DM config errors
  - "UNRECONCILED resources: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false"
  - "2022-10-31T11:58:14.047077043Z\tERROR\tcontroller.host\tuser data error\t{\"request\": \"deployment/controller-0\", \"error\": \"unable to find interface UUID for port: enp0s4\"}" 
FAILED TASK: TASK [Output dm logs]  Wednesday 28 September 2022  02:45:13 +0000 (0:00:00.065)       0:03:28.828   : PlaybookExecutionFailed: Playbook execution failed, command=['ansible-playbook', u'/opt/platform/deploy/22.12/deploy_playbook_wind-river-cloud-platform-deployment-manager.yaml', '-e', u'@/var/opt/dc/ansible/subcloud1_deploy_values.yml', '-i', u'/var/opt/dc/ansible/subcloud1_inventory.yml', '--limit', u'subcloud1']


- Ansible playbook log:

fatal: [subcloud1]: FAILED! => changed=true 
  attempts: 40
  cmd: |-
    source /etc/platform/openrc;
     system host-show "controller-0" | awk '$2 ~ /^task/ {print $4}'
  delta: '0:00:02.071616'
  end: '2022-10-31 12:00:37.518037'
  rc: 0
  start: '2022-10-31 12:00:35.446421'
  stderr: ''
  stderr_lines: []
  stdout: '|'
  stdout_lines: <omitted>
...ignoring

TASK [Show waiting unlock] *****************************************************
Wednesday 28 September 2022  02:45:11 +0000 (0:02:38.743)       0:03:27.251 *** 
ok: [subcloud1] => 
  msg:
  - 'waiting: |'
  - 'waiting: '

TASK [Retrieve kubectl resources reconciled status] ****************************
Wednesday 28 September 2022  02:45:11 +0000 (0:00:00.056)       0:03:27.308 *** 
changed: [subcloud1]

TASK [Show unreconciled resources] *********************************************
Wednesday 28 September 2022  02:45:12 +0000 (0:00:00.502)       0:03:27.811 *** 
ok: [subcloud1] => 
  msg:
  - 'recon: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false'

TASK [Get dm pod name] *********************************************************
Wednesday 28 September 2022  02:45:12 +0000 (0:00:00.054)       0:03:27.865 *** 
changed: [subcloud1]

TASK [Retrieve dm logs] ********************************************************
Wednesday 28 September 2022  02:45:12 +0000 (0:00:00.408)       0:03:28.274 *** 
changed: [subcloud1]

TASK [Show logs] ***************************************************************
Wednesday 28 September 2022  02:45:13 +0000 (0:00:00.488)       0:03:28.762 *** 
ok: [subcloud1] => 
  msg:
  - "Pod log: 2022-10-31T11:58:14.047077043Z\tERROR\tcontroller.host\tuser data error\t{\"request\": \"deployment/controller-0\", \"error\": \"unable to find interface UUID for port: enp0s4\"}"
  - 'err: '

TASK [Waiting for trigger unlock.] **********************************************************
Wednesday 28 September 2022  02:45:13 +0000 (0:00:00.065)       0:03:28.828 *** 
fatal: [subcloud1]: FAILED! => changed=false 
  msg:
  - Fail waiting for host unlock to be triggered. It could be due to DM config errors
  - 'UNRECONCILED resources: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false'
  - "2022-10-31T11:58:14.047077043Z\tERROR\tcontroller.host\tuser data error\t{\"request\": \"deployment/controller-0\", \"error\": \"unable to find interface UUID for port: enp0s4\"}"

PLAY RECAP *********************************************************************
subcloud1                  : ok=41   changed=26   unreachable=0    failed=1



---------------------------------------------------------------------------------------------------


* Case 3: Error in controller-1

In this case, procedure to test it is the same, but editing interface name
into controller1 hostprofile.
After that, run reconfigure command.


output:

[sysadmin@controller-0 ~(keystone_admin)]$ dcmanager subcloud list
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| id | name      | management | availability | deploy status | sync    | backup status | backup datetime |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
| 11 | subcloud1 | unmanaged  | offline      | complete      | unknown | None          | None            |
+----+-----------+------------+--------------+---------------+---------+---------------+-----------------+
[sysadmin@controller-0 ~(keystone_admin)]$ dcmanager subcloud errors 11
No errors present

- /var/log/dcmanager/dcmanager.log:

TASK [Show waiting unlock] *****************************************************
Wednesday 28 September 2022  05:12:18 +0000 (0:01:32.025)       0:02:18.975 *** 
ok: [subcloud1] => 
  msg:
  - 'waiting: Unlocking'
  - 'waiting: '

TASK [Retrieve kubectl resources reconciled status] ****************************
Wednesday 28 September 2022  05:12:18 +0000 (0:00:00.051)       0:02:19.027 *** 
changed: [subcloud1]

TASK [Show unreconciled resources] *********************************************
Wednesday 28 September 2022  05:12:18 +0000 (0:00:00.530)       0:02:19.558 *** 
ok: [subcloud1] => 
  msg:
  - 'recon: host.starlingx.windriver.com/controller-0   locked           disabled      online         controller0-profile   false    false'

TASK [Get dm pod name] *********************************************************
Wednesday 28 September 2022  05:12:18 +0000 (0:00:00.076)       0:02:19.634 *** 
changed: [subcloud1]

TASK [Retrieve dm logs] ********************************************************
Wednesday 28 September 2022  05:12:19 +0000 (0:00:00.561)       0:02:20.195 *** 
changed: [subcloud1]

TASK [Show logs] ***************************************************************
Wednesday 28 September 2022  05:12:20 +0000 (0:00:00.828)       0:02:21.024 *** 
ok: [subcloud1] => 
  msg:
  - 'Pod log: '
  - 'err: '

TASK [Waiting for trigger unlock.] *********************************************
Wednesday 28 September 2022  05:12:20 +0000 (0:00:00.071)       0:02:21.095 *** 
skipping: [subcloud1]

PLAY RECAP *********************************************************************
subcloud1                  : ok=41   changed=26   unreachable=0    failed=0


- Ansible playbook log:

2022-09-28 05:09:58.338 1550496 INFO dcmanager.manager.service [req-421621d9-1c60-445a-887c-8882fc50d089 f0486ee592ab4e02bf039cd78048629e - - default default] Handling reconfigure_subcloud request for: 11
2022-09-28 05:09:58.339 1550496 INFO dcmanager.manager.subcloud_manager [req-421621d9-1c60-445a-887c-8882fc50d089 f0486ee592ab4e02bf039cd78048629e - - default default] Reconfiguring subcloud 11.
2022-09-28 05:09:58.381 1550496 INFO dcmanager.manager.subcloud_manager [-] Starting deploy of subcloud1
2022-09-28 05:12:20.561 1550496 INFO dcmanager.manager.subcloud_manager [-] Successfully deployed subcloud1

ttakenak reviewed (approved):
I agree this change since current target would be checking error until starting unlock in controller-0.

yjian118 commented on docs/playbooks/wind-river-cloud-platform-deployment-manager.yaml:
```diff
@@ -284,3 +305,83 @@
+            system host-show "{{ get_host_name.stdout}}" | awk '$2 ~ /^task/ {print $4}'
+          register: get_show_task_status
+          until: ("Unlocking" in get_show_task_status.stdout)
+          retries: 60
```
As the default delay is 5s, here we wait at most 5 min, which should be long enough
